msal/application.py

```diff
@@ -260,6 +301,27 @@ def _build_telemetry_context(
             self._telemetry_buffer, self._telemetry_lock, api_id,
             correlation_id=correlation_id, refresh_reason=refresh_reason)
 
+    def _get_regional_authority(self, central_authority):
+        is_region_specified = bool(self._region_configured
+            and self._region_configured != self.ATTEMPT_REGION_DISCOVERY)
+        self._region_detected = self._region_detected or _detect_region(
+            self.http_client if self._region_configured is not None else None)
+        if (is_region_specified and self._region_configured != self._region_detected):
+            logger.warning('Region configured ({}) != region detected ({})'.format(
+                repr(self._region_configured), repr(self._region_detected)))
+        region_to_use = (
+            self._region_configured if is_region_specified else self._region_detected)
+        if region_to_use:
+            logger.info('Region to be used: {}'.format(repr(region_to_use)))
+            regional_host = ("{}.login.microsoft.com".format(region_to_use)
```

Comment:
I think this does not take into account sovereign and air-gapped clouds. If the original env is the public one (i.e. "login.microsoftonline.com", "login.windows.net", "login.microsoft.com" or "sts.windows.net" ... these are the only ones we know about), then use `<region>.login.microsoft.com`

Otherwise, use `<region>.<original_env>`, e.g. if the original env is "login.microsoftonline.de", then you'd use `<region>.login.microsoftonline.de`. 

This is very important, because app developers write code once and expect it to work on public + sovereign.

Reply:
Note: The specs above has been changed. The latest specs is in this [Microsoft internal doc](https://identitydivision.visualstudio.com/DevEx/_git/AuthLibrariesApiReview?path=/PinAuthToRegion/AAD%20SDK%20Proposal%20to%20Pin%20Auth%20to%20region.md&anchor=algorithm&_a=preview)